ethexe/observer/src/query.rs

```diff
@@ -39,74 +45,207 @@ impl Query {
+        from_block: u32,
+        to_block: u32,
+    ) -> Result<(Vec<H256>, Vec<H256>)> {
+        let total_blocks = to_block - from_block + 1;
+        log::info!(
+            "Starting to load {} blocks from {} to {}",
+            total_blocks,
+            from_block,
+            to_block
+        );
+
+        let fetches = (from_block..=to_block).map(|block_number| {
+            let provider = self.provider.clone();
+            let database = Arc::clone(&self.database);
+            tokio::spawn(async move {
+                let block = provider
+                    .get_block_by_number(BlockNumberOrTag::Number(block_number as u64), false)
+                    .await?;
+                let block = block
+                    .ok_or_else(|| anyhow!("Block not found for block number {}", block_number))?;
+
+                let height = u32::try_from(block.header.number.ok_or_else(|| {
+                    anyhow!(
+                        "Block number not found for block hash {}",
+                        block.header.hash.unwrap()
+                    )
+                })?)
+                .map_err(|err| anyhow!("Ethereum block number not fit in u32: {}", err))?;
+                let timestamp = block.header.timestamp;
+                let block_hash = H256(block.header.hash.unwrap().0);
+                let parent_hash = H256(block.header.parent_hash.0);
 
-        // First we need to find the nearest valid block
-        let mut hash = block_hash;
-        loop {
-            // TODO: limit deepness
+                let header = BlockHeader {
+                    height,
+                    timestamp,
+                    parent_hash,
+                };
 
-            if hash == self.genesis_block_hash {
-                // Genesis block is always valid
-                log::trace!("Genesis block reached: {hash}");
-                break;
-            }
+                database.set_block_header(block_hash, header);
+
+                Ok::<H256, anyhow::Error>(block_hash)
+            })
+        });
+
+        // Fetch events in block range.
+        let mut blocks_events =
+            read_block_events_batch(from_block, to_block, &self.provider, self.router_address)
+                .await?;
+
+        // Collect results
+        let mut block_hashes = Vec::new();
+        let mut committed_blocks = Vec::new();
+        for result in futures::future::join_all(fetches).await {
+            let block_hash = result??;
+            // Set block events, empty vec if no events.
+            self.database.set_block_events(
+                block_hash,
+                blocks_events.remove(&block_hash).unwrap_or_default(),
+            );
+            committed_blocks.extend(self.get_committed_blocks(block_hash).await?);
+            block_hashes.push(block_hash);
+        }
+        block_hashes.reverse();
+        log::trace!("{} blocks loaded", block_hashes.len());
+
+        Ok((block_hashes, committed_blocks))
```

Comment:
`block_hashes` was collected using block numbers which is not safe. 

More safe solution is to set data in database for all blocks `from_number..to_number` and that __ALL__ where we use numbers - only to access data for blocks.

After that the same logic from load_chain must be used. Because we have already loaded all blocks it will be very fast.

Reply:
Solved